### ISSUE TITLE
Migrating Nx core to Rust & deprecating custom runners

### ISSUE THREAD

**Issue by vsavkin** [nrwl/nx#28434]
Hi folks,

We are migrating Nx core from TypeScript to Rust, with a target completion date of 2025. The main reasons are to improve Nx's speed, reduce its package size, and enable a more powerful command-line interface.

While the rewrite to Rust brings many benefits, it also introduces some limitations, particularly with extensibility APIs. One example is the rarely-used API that allowed users to completely swap out Nx's task runner. While this offered theoretical "infinite" flexibility, it also meant Nx couldn't provide direct support for the use cases it enabled, leading us to discourage its use. In fact, we stopped generating any task runners properties in new Nx workspaces several years ago.

Despite the flexibility, very few concrete use cases for custom task runners emerged. The most common was larger teams wanting to use a self-hosted remote cache. Many such teams now use Nx Cloud (which starts free for small teams), but some still value the ability to choose their own solution.

To address this, we've introduced Nx Powerpack, which includes a dedicated solution for remote caching. Powerpack users can opt for our S3 or shared filesystem cache via new Nx plugins. If your team needs additional plugins for remote caching, please contact us at powerpack-support@nrwl.io.

Another niche use of custom task runners was allowing Nx contributors to experiment with runtimes like Bun and Deno before they were officially supported. We encourage contributors to work with us directly to integrate such changes into Nx's open-source task runner.

We understand that this change would mean that some teams may need Nx Powerpack.  While it's a premium product designed for professional teams, we know that every Nx user's situation is unique. We therefore have the following options available:

- Open-source workspaces can use Nx Powerpack and its plugins for free, just like they have always been able to use Nx Cloud's Pro plan for free.
- If you belong to a smaller team with limited resources but feel you would benefit from the features of Nx Powerpack, please contact us at powerpack-support@nrwl.io, and we will help you out.

Finally, if you have specific use cases for custom task runners that we haven't covered, don't hesitate to open an issue. 

Thank you,
Victor Savkin & Nx Core Team

**Comment by Jordan-Hall** [comment]
> Another niche use of custom task runners was allowing Nx contributors to experiment with runtimes like Bun and Deno before they were officially supported. We encourage contributors to work with us directly to integrate such changes into Nx's open-source task runner.

With regard to this, I would have helped add support for Bun into Nx's own task runner, but at the time Bun didn't support half the modules Nx uses and would only work with web workers over IPC. I did actually provide feedback to the Nx team as there were some benefits over IPC.

I want to express my disappointment that Nx has refused to consider opening up remote caching in Rust. All competitors have either opened up this ability or are working on it.

1. https://rushjs.io/pages/maintainer/build_cache/#enabling-cloud-storage
2. https://www.pantsbuild.org/2.21/docs/using-pants/remote-caching-and-execution
3. https://bazel.build/remote/caching
4. https://turbo.build/repo/docs/core-concepts/remote-caching#self-hosting
5. https://github.com/moonrepo/moon/issues/1520#issuecomment-2347311851

While I respect that Nx has venture capital to consider and needs a return on investment, and that this is Nrwl's business, I believe most enterprises would still pay for Powerpack. However, smaller teams and indie developers would much prefer to host their own. Personally, I have privacy concerns with any cloud—no offence—and would much prefer to control it myself. Taking away a functionality that you admitted was used by the community and putting it behind a paywall is disheartening.

Regarding Powerpack, I find the pricing confusing, particularly in terms of how many seats are required. Does each developer need a seat to use Powerpack? Have you considered selling sections of Powerpack independently and adjusting the pricing for small teams?

**Comment by Elyahou** [comment]
Is it planned to expose an API to add custom remote caching in the Rust implementation ?

**Comment by Jordan-Hall** [comment]
> * If you belong to a smaller team with limited resources but feel you would benefit from the features of Nx Powerpack, please contact us at [powerpack-support@nrwl.io](mailto:powerpack-support@nrwl.io), and we will help you out.


Sorry not to keep hating but that comment feels belittling. I'm sorry you made this amazing tool. Caching was extremely useful before but we have limited funds as we not backed by VC yet. Please  give me charity.


honestly thats how it reads

**Comment by ndrsg** [comment]
Shared cache Is also necessary to reuse cache between pipeline runs in a stateless build cluster like kubernetes or will there be another option to achieve this.

From my understanding, without Powerpack or NX Cloud it will not be possible to have "Smart Monorepos AND fast CI" if you dont have a stateful CI executor, but thats something that NX wanted to provide, isnt it?

If you allow a suggestion, you may put features like special/ complicated executors or plugins for e.g. Cmake, Conan or other C/C++ stuff in a powepack, but not core features. 

I mean 20% faster sounds great at first, but compared to what? Typical build/test task might take some minutes, it does not really matter if it takes 100ms or 80ms to load it from cache instead.

**Comment by el-davo** [comment]
In my current job we have a team of 20 developers working on a massive monorepo with NX as the underlying framework. 

We chose NX specifically for its free tier features including on prem caching as security is paramount to our organization. We use on prem minio s3 storage that we control and administer ourselves. 

I was surprised to learn this week by means of a warning when run `pnpm nx serve` that on prem caching will soon be behind a paywall

![Image](https://github.com/user-attachments/assets/154b71fd-4ed7-4faf-b2f9-a58861fd12ba)

To learn such an important update from a random warning is really poor communication from NX.

I had a look at site for activating powerpack and with the amount of developers working on our project the total comes up to $495.00
USD / MONTH. Again we administer our own storage. Why would we pay such a massive amount every month for this? Does anyone think this is acceptable? And what do we get in return? A rewrite in rust that noone asked for? 

A 20% speed increase in a command starting up is nothing in comparison to the speed increase that remote caching offers. 

Unfortunatly NX has now lost credibility within our organization, reason being, if you can randomly put caching behind a paywall what's to stop NX from putting other features behind a paywall?

- Want to build an application with module federation - sorry that will be an extra $20 a month
- Want to use the task scheduler - sorry that will be an extra $20 a month
- Want to use local caching - Purchase nx mega powerpack to unlock

We have already started planning for migrating off of NX and are looking into turbo repo as we are not happy with essentially being held to ransom for trusting NX.

I should note that personally I was a massive fan of NX, I was the one who incorporated it into our development process and honestly it has been an amazing journey. I even wrote an article about it here - https://medium.com/better-programming/migrating-an-angular-application-to-module-federation-with-nx-74f7664bab3e?source=your_stories_page-------------------------------------

 Thank you for all the work up until now, but if this change goes ahead we will be moving on and I suspect a lot of others will also

**Comment by Jordan-Hall** [comment]
> We have already started planning for migrating off of NX and are looking into turbo repo as we are not happy with essentially being held to ransom for trusting NX.
> 
> Thank you for all the work up until now, but if this change goes ahead we will be moving on and I suspect a lot of others will also

100% agree, I've stopped making NX a priority for client. I'll be recommending other tools as well going forward. NX to me now has zero credibility. I also found this quiet disingenuous too around codeowner.

https://github.com/nrwl/nx/issues/20926

> This is something we've discussed, but is not a priority currently. The best solution for now is to create a local plugin.

He went on to create a community plugin

npm i @swapniltech0390/nx-codeowners

https://www.npmjs.com/package/@swapniltech0390/nx-codeowners

**Comment by Jordan-Hall** [comment]
The custom NX license states:

> “Nx IP” means the Software, algorithms, technology, databases, tools, know-how, or processes used to provide or deliver the Software or related services, and its documentation (“Documentation”), including all improvements, modifications, or derivative works (regardless of authorship), and all intellectual property rights (“IPR”) in any of the foregoing


However, much of this appears to be documented and available under the MIT version of the code. How is this license valid if the API/know-how for producing related software comes from the MIT version?

Does this mean any indie hacker using APIs provided by the MIT or any tweaks to the MIT code to open up would be a breach.

**Comment by Ligrev861** [comment]
I'm going to post this from an anonymous account because I, like you, am a walking and talking bag of emotions, and I want our conversation in threads and over the phone to have their own context. I will try to be unbiased and I'll probably fail.

There are several issues here.

1. The deprecation that prompted this series of threads
I understand Nx wants to be commercially successful but commercial viability looks challenged.

Nx Cloud offers remote caching, agents, dashboards, and log visualization. The paid version even feeds this data into an AI to generate basic insights and costs $250 for 30 people.

Nx Powerpack offers a JSON-to-CODEOWNERS converter, an ESLint rule, and cache self-hosting which was previously available but not marketed.
All current threads on GitHub talk about Powerpack in the context of caching and custom runners so it's not a stretch to assume that caching is its primary selling point.
It offers that at the cost of $717.50 for 30 people.

Assuming an organization with 30 developers, that's almost triple the price if you want to use your own cache and if you gave up on agents, dashboards, and log viz.

I understand these are two completely different products aimed at different target groups but as an engineer I can't help but compare the amount of effort that has gone in one vs. the other and the two are not close. And the way these are perceived informs whether engineers would recommend and implement Nx in their organization.

2. Unclear terminology in sales pages
Nx would not be my first example for a great documentation, but the [Nx pricing](https://nx.dev/pricing) talks about "contributors" while the [Powerpack pricing](https://nx.dev/powerpack) talks about "seats".
Are those the same? What counts as a contributor? If a backender who doesn't know Nx is in the tech stack runs a command that sets up their environment and that command happens to run Nx in the background, does that constitue as a seat, a contributor, or neither?

3. Unclear licensing
Jordan Hall raises an excellent question in [his comment](https://github.com/nrwl/nx/issues/28434#issuecomment-2413484871), I won't repeat it

4. Removing previously available functionality
I understand the arguments about making Nx more responsive in benchmarks. Nx has come a very long way and is an excellent piece of software. We all want to see it become the best it can be, but no one wants to part ways with impactful features. We can be generous with phrasing but there's no changing the fact that an impactful feature is getting removed and its primary use is sent into closed-source land.  
People are reconsidering their relationship with Nx over this. This is an extremely bad change in terms of PR and it's really tempting to describe it as a rug-pull. Trust is lost all too easily.

5. Comparison with competing products
At some point, usually before buying anything, people will compare their options. Nx charges for CI optimization. The biggest differentiator there is the cache. Other products listed in [Jordan's other comment](https://github.com/nrwl/nx/issues/28434#issuecomment-2411948721) offer self-hosted caching for free.
There are other things that make Nx great but this is the big one, and Nx is behind.

7. Public communication: correctness and timeliness
The "[Evolving Nx](https://nx.dev/blog/evolving-nx)" post spends a lot of time trying to convince developers that no functionality is getting removed (which is disingenuous) and that people should get paid (which no one argues against). These two arguments put next to each other sound manipulative. The way this post is phrased makes it clear that Nx expected the backlash which only makes things worse.

Timeliness is another thing. The post is dated Sep 25th and its follow-up explanation is dated Oct 14th.

6. Public communication: subjective interpretation
> If you belong to a smaller team with limited resources but feel you would benefit from the features of Nx Powerpack, please contact us at powerpack-support@nrwl.io, and we will help you out.

This is great, except that now Nx now has a history of removing core features. This reads as "we can help you adopt Nx but you'll have to hope we won't change the terms again".

> Everything in Powerpack is new functionality, not previously free features that we’re now putting behind a paywall

This is only "technically correct" as evident by the recent issues and PRs getting created around this. Nx is offering a product to developers. Nx is taking away building blocks from developers. Nx now has to somehow convince their main audience that this is a good thing.

My team is prototyping micro frontends via module federation and NX in our company and we are potentially exploring the enterprise version of Nx. Our frontend team has always been lean so this wouldn't be a hard sell, nor it would be my money - but the actions and the communication of Nx around this are very encouraging to look for alterntaives.
This all needs addressing - and ideally a step back.

**Comment by Ligrev861** [comment]
I'm going to post this from an anonymous account because I, like you, am a walking and talking bag of emotions, and I want our conversation in threads and over the phone to have their own context. I will try to be unbiased and I'll probably fail.

There are several issues here.

**The deprecation that prompted this series of threads**
I understand Nx wants to be commercially successful but the recent advances in commercial viability look challenged.  
Nx Cloud offers remote caching, agents, dashboards, and log visualization. The paid version even feeds this data into an AI to generate basic insights and costs $250 for 30 people.  

Nx Powerpack offers a JSON-to-CODEOWNERS converter, an ESLint rule, and cache self-hosting which was previously available but not marketed.
  
All current threads on GitHub talk about Powerpack in the context of caching and custom runners so it's not a stretch to assume that caching is its primary selling point.
It offers that at the cost of $717.50 for 30 people.

Assuming an organization with 30 developers, that's almost triple the price if you want to use your own cache and if you gave up on agents, dashboards, and log viz.

I understand these are two completely different products aimed at different target groups but as an engineer I can't help but compare the amount of effort that has gone in one vs. the other and the two are not close. And the way these are perceived informs whether engineers would recommend and implement Nx in their organization.

2. Unclear terminology in sales pages
Nx would not be my first example for a great documentation, but the [Nx pricing](https://nx.dev/pricing) talks about "contributors" while the [Powerpack pricing](https://nx.dev/powerpack) talks about "seats".
Are those the same? What counts as a contributor? If a backender who doesn't know Nx is in the tech stack runs a command that sets up their environment and that command happens to run Nx in the background, does that constitue as a seat, a contributor, or neither?

3. Unclear licensing
Jordan Hall raises an excellent question in [his comment](https://github.com/nrwl/nx/issues/28434#issuecomment-2413484871), I won't repeat it

4. Removing previously available functionality
I understand the arguments about making Nx more responsive in benchmarks. Nx has come a very long way and is an excellent piece of software. We all want to see it become the best it can be, but no one wants to part ways with impactful features. We can be generous with phrasing but there's no changing the fact that an impactful feature is getting removed and its primary use is sent into closed-source land.  
People are reconsidering their relationship with Nx over this. This is an extremely bad change in terms of PR and it's really tempting to describe it as a rug-pull. Trust is lost all too easily.

5. Comparison with competing products
At some point, usually before buying anything, people will compare their options. Nx charges for CI optimization. The biggest differentiator there is the cache. Other products listed in [Jordan's other comment](https://github.com/nrwl/nx/issues/28434#issuecomment-2411948721) offer self-hosted caching for free.
There are other things that make Nx great but this is the big one, and Nx is behind.

7. Public communication: correctness and timeliness
The "[Evolving Nx](https://nx.dev/blog/evolving-nx)" post spends a lot of time trying to convince developers that no functionality is getting removed (which is disingenuous) and that people should get paid (which no one argues against). These two arguments put next to each other sound manipulative. The way this post is phrased makes it clear that Nx expected the backlash which only makes things worse.

Timeliness is another thing. The post is dated Sep 25th and its follow-up explanation is dated Oct 14th.

6. Public communication: subjective interpretation
> If you belong to a smaller team with limited resources but feel you would benefit from the features of Nx Powerpack, please contact us at powerpack-support@nrwl.io, and we will help you out.

This is great, except that now Nx now has a history of removing core features. This reads as "we can help you adopt Nx but you'll have to hope we won't change the terms again".

> Everything in Powerpack is new functionality, not previously free features that we’re now putting behind a paywall

This is only "technically correct" as evident by the recent issues and PRs getting created around this. Nx is offering a product to developers. Nx is taking away building blocks from developers. Nx now has to somehow convince their main audience that this is a good thing.

My team is prototyping micro frontends via module federation and NX in our company and we are potentially exploring the enterprise version of Nx. Our frontend team has always been lean so this wouldn't be a hard sell, nor it would be my money - but the actions and the communication of Nx around this are very encouraging to look for alterntaives.
This all needs addressing - and ideally a step back.

**Comment by Ligrev861** [comment]
I'm going to post this from an anonymous account because I, like you, am a walking and talking bag of emotions, and I want our conversation in threads and over the phone to have their own context. I will try to be unbiased and I'll probably fail.

There are several issues here.

**The deprecation that prompted this series of threads**
I understand Nx wants to be commercially successful but commercial viability looks challenged.  
Nx Cloud offers remote caching, agents, dashboards, and log visualization. The paid version even feeds this data into an AI to generate basic insights and costs $250 for 30 people.

Nx Powerpack offers a JSON-to-CODEOWNERS converter, an ESLint rule, and cache self-hosting which was previously available but not marketed.
All current threads on GitHub talk about Powerpack in the context of caching and custom runners so it's not a stretch to assume that caching is its primary selling point.
It offers that at the cost of $717.50 for 30 people.

Assuming an organization with 30 developers, that's almost triple the price if you want to use your own cache and if you gave up on agents, dashboards, and log viz.

I understand these are two completely different products aimed at different target groups but as an engineer I can't help but compare the amount of effort that has gone in one vs. the other and the two are not close. And the way these are perceived informs whether engineers would recommend and implement Nx in their organization.

**Unclear terminology in sales pages**
Nx would not be my first example for a great documentation, but the [Nx pricing](https://nx.dev/pricing) talks about "contributors" while the [Powerpack pricing](https://nx.dev/powerpack) talks about "seats".
Are those the same? What counts as a contributor? If a backender who doesn't know Nx is in the tech stack runs a command that sets up their environment and that command happens to run Nx in the background, does that constitue as a seat, a contributor, or neither?

**Unclear licensing**
Jordan Hall raises an excellent question in [his comment](https://github.com/nrwl/nx/issues/28434#issuecomment-2413484871), I won't repeat it

**Removing previously available functionality**
I understand the arguments about making Nx more responsive in benchmarks. Nx has come a very long way and is an excellent piece of software. We all want to see it become the best it can be, but no one wants to part ways with impactful features. We can be generous with phrasing but there's no changing the fact that an impactful feature is getting removed and its primary use is sent into closed-source land.  
People are reconsidering their relationship with Nx over this. This is an extremely bad change in terms of PR and it's really tempting to describe it as a rug-pull. Trust is lost all too easily.

**Comparison with competing products**
At some point, usually before buying anything, people will compare their options. Nx charges for CI optimization. The biggest differentiator there is the cache. Other products listed in [Jordan's other comment](https://github.com/nrwl/nx/issues/28434#issuecomment-2411948721) offer self-hosted caching for free.
There are other things that make Nx great but this is the big one, and Nx is behind.

**Public communication: correctness and timeliness**
The "[Evolving Nx](https://nx.dev/blog/evolving-nx)" post spends a lot of time trying to convince developers that no functionality is getting removed (which is disingenuous) and that people should get paid (which no one argues against). These two arguments put next to each other sound manipulative. The way this post is phrased makes it clear that Nx expected the backlash which only makes things worse.

Timeliness is another thing. The post is dated Sep 25th and its follow-up explanation is dated Oct 14th.

**Public communication: subjective interpretation**
> If you belong to a smaller team with limited resources but feel you would benefit from the features of Nx Powerpack, please contact us at powerpack-support@nrwl.io, and we will help you out.

This is great, except that now Nx now has a history of removing core features. This reads as "we can help you adopt Nx but you'll have to hope we won't change the terms again".

> Everything in Powerpack is new functionality, not previously free features that we’re now putting behind a paywall

This is only "technically correct" as evident by the recent issues and PRs getting created around this. Nx is offering a product to developers. Nx is taking away building blocks from developers. Nx now has to somehow convince their main audience that this is a good thing.

My team is prototyping micro frontends via module federation and NX in our company and we are potentially exploring the enterprise version of Nx. Our frontend team has always been lean so this wouldn't be a hard sell, nor it would be my money - but the actions and the communication of Nx around this are very encouraging to look for alterntaives.
This all needs addressing - and ideally a step back.

**Comment by Ligrev861** [comment]
I'm going to post this from an anonymous account because our threads and phone calls should not be about this.
There are several issues here.

**The deprecation that prompted this series of threads**
I understand Nx wants to be commercially successful but commercial viability looks challenged.  
Nx Cloud offers remote caching, agents, dashboards, and log visualization. The paid version even feeds this data into an AI to generate basic insights and costs $250 for 30 people.

Nx Powerpack offers a JSON-to-CODEOWNERS converter, an ESLint rule, and cache self-hosting which was previously available but not marketed.
All current threads on GitHub talk about Powerpack in the context of caching and custom runners so it's not a stretch to assume that caching is its primary selling point.
It offers that at the cost of $717.50 for 30 people.

Assuming an organization with 30 developers, that's almost triple the price if you want to use your own cache and if you gave up on agents, dashboards, and log viz.

I understand these are two completely different products aimed at different target groups but as an engineer I can't help but compare the amount of effort that has gone in one vs. the other and the two are not close. And the way these are perceived informs whether engineers would recommend and implement Nx in their organization.

**Unclear terminology in sales pages**
Nx would not be my first example for a great documentation, but the [Nx pricing](https://nx.dev/pricing) talks about "contributors" while the [Powerpack pricing](https://nx.dev/powerpack) talks about "seats".
Are those the same? What counts as a contributor? If a backender who doesn't know Nx is in the tech stack runs a command that sets up their environment and that command happens to run Nx in the background, does that constitue as a seat, a contributor, or neither?

**Unclear licensing**
Jordan Hall raises an excellent question in [his comment](https://github.com/nrwl/nx/issues/28434#issuecomment-2413484871), I won't repeat it

**Removing previously available functionality**
I understand the arguments about making Nx more responsive in benchmarks. Nx has come a very long way and is an excellent piece of software. We all want to see it become the best it can be, but no one wants to part ways with impactful features. We can be generous with phrasing but there's no changing the fact that an impactful feature is getting removed and its primary use is sent into closed-source land.  
People are reconsidering their relationship with Nx over this. This is an extremely bad change in terms of PR and it's really tempting to describe it as a rug-pull. Trust is lost all too easily.

**Comparison with competing products**
At some point, usually before buying anything, people will compare their options. Nx charges for CI optimization. The biggest differentiator there is the cache. Other products listed in [Jordan's other comment](https://github.com/nrwl/nx/issues/28434#issuecomment-2411948721) offer self-hosted caching for free.
There are other things that make Nx great but this is the big one, and Nx is behind.

**Public communication: correctness and timeliness**
The "[Evolving Nx](https://nx.dev/blog/evolving-nx)" post spends a lot of time trying to convince developers that no functionality is getting removed (which is disingenuous) and that people should get paid (which no one argues against). These two arguments put next to each other sound manipulative. The way this post is phrased makes it clear that Nx expected the backlash which only makes things worse.

Timeliness is another thing. The post is dated Sep 25th and its follow-up explanation is dated Oct 14th.

**Public communication: subjective interpretation**
> If you belong to a smaller team with limited resources but feel you would benefit from the features of Nx Powerpack, please contact us at powerpack-support@nrwl.io, and we will help you out.

This is great, except that now Nx now has a history of removing core features. This reads as "we can help you adopt Nx but you'll have to hope we won't change the terms again".

> Everything in Powerpack is new functionality, not previously free features that we’re now putting behind a paywall

This is only "technically correct" as evident by the recent issues and PRs getting created around this. Nx is offering a product to developers. Nx is taking away building blocks from developers. Nx now has to somehow convince their main audience that this is a good thing.

My team is prototyping micro frontends via module federation and NX in our company and we are potentially exploring the enterprise version of Nx. Our frontend team has always been lean so this wouldn't be a hard sell, nor it would be my money - but the actions and the communication of Nx around this are very encouraging to look for alterntaives.
This all needs addressing - and ideally a step back.

**Comment by JCMais** [comment]
heh I was looking into NX, but the lack of self-hosted caching without having to pay for an enterprise license is a no-go. I found this thread by looking into alternatives. NX Cloud sounds good enough, but even if we decided to pay, just not having the option to do the remote caching in-house, and for free, is a huge blocker.

**Comment by Jordan-Hall** [comment]
I'm encourage by the new license thats been recently added 
![Image](https://github.com/user-attachments/assets/2c740016-2e93-4ea3-a924-82f1d0906f84)

This at least address concerns it felt like we need to beg for access. However their seem to be little documentation on what "restricted license " means compared to the unrestricted.

@vsavkin  is their a link or something  we can read to see what the restriction is

**Comment by pete-mcwilliams** [comment]
looks to be very restrictive:
Single workspace license type
1 seat

**Comment by Jordan-Hall** [comment]
> looks to be very restrictive: Single workspace license type 1 seat

Why i'm asking because it looks no different for me
![Image](https://github.com/user-attachments/assets/82846ee1-6b13-4e8e-b114-4d965e80b108)

**Comment by pete-mcwilliams** [comment]
This is what I see
![Image](https://github.com/user-attachments/assets/8905aeae-bdfd-485d-964d-e2e6f57c360c)

**Comment by Jordan-Hall** [comment]
You need to change the Number of developers

**Comment by pete-mcwilliams** [comment]
Then it is not restricted at all 👍

**Comment by juristr** [comment]
Sorry for the delayed response, we're still catching up after Monorepo World.

I won't repeat what Victor [mentioned in his post](#issue-2586704806), so please check that out first. However, I want to focus on a couple of things:
- Our goal has never been to block anyone or take anything away. We have a **strong history of supporting OSS** at both the company and individual level, and we continue in that support by offering our paid products for free to OSS projects (e.g., Nx Cloud is free for OSS).
- That said, we sought to create a product for large enterprises to help us in sustaining our open-source efforts. This isn't about VC funding; it's about transitioning from a consulting business to a product company, as Jeff outlined in his blog post, and making open-source work sustainable.

We're grateful for the feedback :pray:. It showed us that while we had mentioned it in a few places, we missed the mark and needed to be much more explicit in our support for smaller teams. We've now updated the [Powerpack page](https://nx.dev/powerpack) and the license application process to make these points more clear:
- We offer **Powerpack for free to OSS projects** ([apply here](https://nx.dev/powerpack/special-offer)); again note that Nx Cloud is [also free for OSS](https://nx.dev/pricing#oss).
- We offer **Powerpack for free to small teams** ([reach out here](https://cloud.nx.app/powerpack/purchase)).

If you're a small team and you rely on custom task runners for S3 caching, please contact us—we definitely don't want anyone feeling blocked by this.

**Comment by vsavkin** [comment]
Thank you so much, Juri. 

Everyone, I apologize for my unclear communication. And I appreciate everyone's constructive feedback. 

My Twitter DMs are open to anyone who wants to talk in good faith. I'm also happy to jump on a call. I chatted with many folks last week and the week before; this remains my top priority. 

Nx is a small business with a few dozen engineers. We know how hard it is to run a small business, so we're offering all small businesses remote cache implementations for free.

**Comment by vsavkin** [comment]
@Jordan-Hall  That is my personal twitter account. The Nx account didn't block you.  

You declined my invitation to talk to you. After I texted with you in good faith, trying to be clear and honest, you published our private communication without my permission, which isn't acting in good faith imo.

That's why I don't think I can trust you, and I decided not to engage with you on social **personally**.  You are free to engage with other folks on the Nx core team though.

I'd also like you to read through https://github.com/nrwl/nx/blob/master/CODE_OF_CONDUCT.md which is there to protect the wellbeing of all Nx contributors, regardless of their position. Further violations of the CoC will leave us with no choice but to restrict your access on GitHub in order to preserve a safe space for our community.

If you want to restart the conversation constructively, I'm happy to have a VC with you.

Thank you!

**Comment by Jordan-Hall** [comment]
> You declined my invitation to talk to you. After I texted with you in good faith, trying to be clear and honest, you published our private communication without my permission, which isn't acting in good faith imo.

I said it was something i didn't like because people seem to be in the dark. It was release as good faith so people know what it was about (not to make you look bad or in bad faith, sorry if you took it that way..

Im sorry that likes things open but sure would take a call now im actually feeling well enough. The new update looks great but seems confusing around "Free restricted license." I cant find documentation on the restrictions,

**Comment by rathpc** [comment]
> _We know how hard it is to run a small business, so we're offering all small businesses remote cache implementations for free._

What is considered a "small team" out of curiosity?

**Comment by el-davo** [comment]
> free for small existing teams

Also what is an existing team?

**Comment by pumano** [comment]
@vsavkin very good idea migrate core to Rust. I also suggest to support rust-based modern toolchain like oxc / oxlint / rolldown / swc / esbuild / rspack / vite etc. Performance matters.

**Comment by comp615** [comment]
> Our focus has been on building additional features that benefit our Enterprise users—like Conformance, Codeowners, and more to come—without impacting smaller teams of developers. We're not looking to impose extra costs on teams that are already bootstrapping their way forward.

It seems like those are great, net new powerpack features. But based on the feedback on this thread, I think everyone would be very happy to see caching provided as free based on the history here and pay for those new functions. Right now it's a bait and switch. And now I have to have a conversation amid cost cutting efforts along the lines of "Hey we need to pony up tens of thousands of dollars just so that thing I committed us to and said was free last year can keep doing excatly what it does"... I'm personally very excited to see where compliance goes. but I don't like being put in that position. It's disingenuous to include caching in: "No, we’re not placing existing features behind a paywall"

We're not a small company, I don't feel legitimized asking for a _broad based exception for powerpack_. I would go to bat and collaborate for these new features. But the money grab on caching needs to be acknowledged; there's no way around it.

**Comment by Jordan-Hall** [comment]
I know there's a real desire to avoid having a JavaScript runtime within Rust code due to performance and package size concerns. However, what about running WebAssembly or dynamically linking with a shared library? It would be great to create plugins without having to use JavaScript. This would be an ideal example to get it working without making it feel like a cash grab.

I'm sure it's a good middle ground, ensuring flexibility while still meeting all your goals.

**Comment by psalv** [comment]
@vsavkin @jeffbcross I think a big piece your team is missing is that for larger companies (like ours) we would be happy to pay for products that add value (which Nx does) but we need an off-ramp and this is what custom task runners/self-hosted caching provided—having these options available to us is a big part of why we chose to integrate with (and pay for) Nx.  

This change has turned your product into a one-way door and I can’t advocate investing in a product that we have no way of pulling out of if our needs change in the future. We try to avoid putting ourselves into a state of vendor lock-in.

**Comment by Jordan-Hall** [comment]
@jeffbcross @vsavkin is it possible to start having RFC for big changes, allow community to provide early feedback ans suggestions. I'm sure people would love to help out more and with a design for making this still work in rust

**Comment by Elyahou** [comment]
![Image](https://github.com/user-attachments/assets/172f9b20-e922-4087-b4bf-dd96ba2d10f6)

You previously claimed that Powerpack was not replacing open-source features with paid subscriptions. However, this blog post change shows otherwise. We’ve now confirmed that functionality—like custom remote caching—has been moved behind a paywall.

It would be fair for the Nx team to remove this limitation and allow the same API to remain accessible in the Rust implementation. If open-source features can simply be reclassified as premium, this sets a dangerous precedent. What’s stopping other Nx APIs from being locked behind future paywalls?

This shift makes relying on Nx uncertain, as it suggests that any feature could be turned into a paid service at any time.

**Comment by ItamarGronich** [comment]
While supporting OSS and small businesses is a noble and just cause you still haven't addressed the concern for big companies yet.

We're a big company.
We chose Nx because of it's amazing free features.
Free remote caching being the a key decision point.
I now have to come to my superiors and say "Hey, Nx now decided to put remote caching behind paywall so we have three options:
1. pay X now
2. give  up remote caching
3. migrate to another solution"

In either way we'll have to pay a large fee - either in buying the nx cloud service or by putting large amount of dev hours on finding another solution, or by giving it up to and eat the extra time remote caching saved us.

In any case, this feel basically like ransom and is not a behavior i would expect from a company such as yours.
Not only that, but you not directly mentioning the pricing change is very disingenuous.
It feels like you're trying to avoid saying that upfront.
You promote this by saying: "HEY! we're moving to RUST! 🦀" But what you should be saying is: "HEY! We now want you to pay for remote caching!"

This kind of behavior trashes the trust and reliability we have in Nx.
Frankly even if this ends up not going through - I'm not sure i would recommend using Nx going forward.
Since at any given time any feature you might be using today could be held ransom in the future.

**Comment by xalechez** [comment]
I'd like to second @ItamarGronich's comment. 

I am currently investigating using Nx in our organisation. We see some huge benefits it could bring us as a business, but as with any new technology we have to go iteratively, starting from early adoption of the OSS, and then, if we see value, attempt to scale it further, considering commercial options like Powerpack or Nx Cloud.

However before considering paying for anything I need to build trust in the tech and the team behind it, so that we don't find ourselves with vendor lock-in in the future. I also need to show the value of it before investing any company's money which I could possibly do with using custom runners (e.g. "hey, here's how we save money on CI bills and eng context switching by using custom cache, how about we buy Nx Cloud and save us the effort of maintaining homegrown solution while also having other huge benefits?"). Now we have a big problem both with trust and the ability to demonstrate the value.

I am still an Nx enthusiast, and I believe that it has huge benefits, but this situation is a **huge red flag** for me.

I haven't reported it back to my superiors yet as I believe that the pushback from community could somehow convince Nx team that their decision, though seeming profitable short-term, can cost them more long-term as trust would be eroded and potential customers who are ready to pay for something that brings them value would have to consider other alternatives, but I will have to do it, and the only answer that comes to mind is to scale back the adoption and consider moving to Turborepo (which was our second choice as it loses to Nx for a number of cases specific to our business).

I appreciate your efforts to make Nx faster by jumping on the web-wide trend of rewriting to Rust, but I would appreciate even more if this decision was scaled back by providing some sort of API for those who would like to build their own caching solution

**Comment by TriPSs** [comment]
We have been using Nx for over 4 years with our [custom runner](https://github.com/TriPSs/nx-extend/tree/master/packages/gcp-task-runner) in all projects, after seeing the warning and trying to investigate what a Powerpack was came to this issue. 

In the past years we already had multiple annoyances of Nx doing things that broke everything for us and we had to spend hours upon hours to figure out what, we accepted this as Nx being Nx because we really liked the ecosystem and the extendability of it. That we now have to pay for something that has always been free is just absurd, even with the "free" licence, why? Why would we now suddeldy need to request and get approved for a license to use what we are/where already using for the last couple of years?

My 2 cents about the migration from Typescript to Rust:
As someone who maintains [multiple plugins](https://github.com/tripss/nx-extend) for Nx (and has never done anything in Rust), I'm not looking forward to this, currently breaking changes for plugins are not always documented, or testing utils that are used internally (#10992) but not available to plugin maintainers I can only see more issues / making it harder to maintain those plugins. Are API's that are now exposed still going to be exposed?

Would rather see more focus on existing tickets/bugs that are now ignored (#22213, #26577) instead of rewriting the core into Rust to improve performance of something that, in my opinion, is not slow.

**Comment by Jordan-Hall** [comment]
@juristr with this bow being closes can we assumed its all going ahead as planned and this should be closes?

**Comment by Elyahou** [comment]
Yes, I don't see any meaningful response from the Nx team on the concerns raised throughout this thread. It would be helpful to understand their approach to these issues, especially to reassure the community about the future direction of Nx. Even if no changes are planned, we still deserve a clear and direct answer.

The current silence only adds to the uncertainty. If APIs and extension points are being removed or moved behind paywalls, users need to know now, so they can make informed decisions about whether Nx will continue to meet their needs or if they should start exploring alternatives. Transparency is key to maintaining trust in the open source community.

**Comment by tidusjar** [comment]
Time to start taking NX out of everything

**Comment by johannesnormannjensen** [comment]
> In my current job we have a team of 20 developers working on a massive monorepo with NX as the underlying framework.
> 
> We chose NX specifically for its free tier features including on prem caching as security is paramount to our organization. We use on prem minio s3 storage that we control and administer ourselves.
> 
> I was surprised to learn this week by means of a warning when run `pnpm nx serve` that on prem caching will soon be behind a paywall
> 
> ![Image](https://github.com/user-attachments/assets/154b71fd-4ed7-4faf-b2f9-a58861fd12ba)
> 
> To learn such an important update from a random warning is really poor communication from NX.
> 
> I had a look at site for activating powerpack and with the amount of developers working on our project the total comes up to $495.00 USD / MONTH. Again we administer our own storage. Why would we pay such a massive amount every month for this? Does anyone think this is acceptable? And what do we get in return? A rewrite in rust that noone asked for?
> 
> A 20% speed increase in a command starting up is nothing in comparison to the speed increase that remote caching offers.
> 
> Unfortunatly NX has now lost credibility within our organization, reason being, if you can randomly put caching behind a paywall what's to stop NX from putting other features behind a paywall?
> 
> * Want to build an application with module federation - sorry that will be an extra $20 a month
> * Want to use the task scheduler - sorry that will be an extra $10 a month
> * Want to use local caching - Purchase nx mega powerpack to unlock
> 
> We have already started planning for migrating off of NX and are looking into turbo repo as we are not happy with essentially being held to ransom for trusting NX.
> 
> I should note that personally I was a massive fan of NX, I was the one who incorporated it into our development process and honestly it has been an amazing journey. I even wrote an article about it here - https://medium.com/better-programming/migrating-an-angular-application-to-module-federation-with-nx-74f7664bab3e?source=your_stories_page-------------------------------------
> 
> Thank you for all the work up until now, but if this change goes ahead we will be moving on and I suspect a lot of others will also

We too use on prem minio storage and would not like to see this changed. It makes sense to pay for the features that NX provides as a service, but for us to pay for using a shared fs cache to setup ourselves as well... What is that about? 

We've been (and myself personall) very happy with NX all through the years, but changes like this makes me rethink what the direction of the whole organisation is moving forward. As others have statet, what will be the next paywall?

**Comment by vsavkin** [comment]
Thank you for engaging with us!

Anyone who has spoken with us in person at conferences knows that the Nx team, myself included, is always happy to discuss what's in progress and what's coming next. However, we want to do a much better job of sharing this in a more public way, so everyone interested can stay informed and have an opportunity to share feedback early on.

To achieve this, we're making a few changes:
1. Publishing a 1-Year Roadmap: We're bringing back roadmaps and revamping the process. This roadmap will outline all major developments for the year, offering a chance to see what's in progress, ask questions, and share concerns. The 2025 roadmap will be published soon.
2. A More Formal RFC Process: For larger API changes (e.g., Crystal, Powerpack are past examples of such changes), we'll pilot a more structured RFC process, providing an opportunity for you to see upcoming changes early and contribute directly to the discussion. We commit to only implementing major changes based on the resolutions of these RFCs, ensuring that your input guides our direction.
3. Hosting Regular Office Hours: We'll host regular office hours, providing a chance for anyone to have direct access to the team.
4. Improving the Nx Champions Program: The Nx Champions program connects 28 prominent community members with the Nx team for feedback and idea exchange. We haven't been using this program as effectively as we'd like. We're going to take action to improve the program such that there's more collaboration, early feedback gathering but also getting their help to spread big ongoing developments better in the community.

I'm personally committed to engaging with the community. If you have questions or concerns, feel free to DM me on Twitter (https://x.com/victorsavkin), and I'll be happy to have a video call with you. You can also engage with Juri in the same fashion (https://x.com/juristr).
Discord is also a great place to engage at.

I'm sorry we didn't do a good enough job of communicating with the community. We are doing our best to address it.

If you have any questions about these updates, please engage with us!

**Comment by Jordan-Hall** [comment]
> 1. Publishing a 1-Year Roadmap: We're bringing back roadmaps and revamping the process. This roadmap will outline all major developments for the year, offering a chance to see what's in progress, ask questions, and share concerns. The 2025 roadmap will be published soon.
> 2. A More Formal RFC Process: For larger API changes (e.g., Crystal, Powerpack are past examples of such changes), we'll pilot a more structured RFC process, providing an opportunity for you to see upcoming changes early and contribute directly to the discussion. We commit to only implementing major changes based on the resolutions of these RFCs, ensuring that your input guides our direction.
> 3. Hosting Regular Office Hours: We'll host regular office hours, providing a chance for anyone to have direct access to the team.
> 4. Improving the Nx Champions Program: The Nx Champions program connects 28 prominent community members with the Nx team for feedback and idea exchange. We haven't been using this program as effectively as we'd like. We're going to take action to improve the program such that there's more collaboration, early feedback gathering but also getting their help to spread big ongoing developments better in the community.



This sounds great, but is it too late for the current approach regarding this issue? Or can you start again and rethink, or allow community input or idea on making a rust base plugin

**Comment by nielskrijger** [comment]
> I'm sorry we didn't do a good enough job of communicating with the community. We are doing our best to address it.

I think you've actually engaged with the community pretty well.

It's the community that is stumped and in disbelief. As said in the opening post;

> We understand that this change would mean that some teams may need Nx Powerpack. While it's a premium product designed for professional teams, we know that every Nx user's situation is unique. 

There should be very little doubt what that means.

For many companies (especially outside of SF or US) the costs for your services are a significant expense for what is essentially some convenience.

In comparison to all the free CI/CD and OS options available, this can be a tough expense to justify.

Obviously; new kids are knocking at the doors. Drastic changes are necessary.

But for those who have contributed to and enjoyed using NX, this feels like an free tool being hijacked by a commercial entity.

Losing some of the community might not impact the bottom line, but without growth, the project/company's future is at risk.

If that growth only comes from imitating competitors, it’s a losing game. Free and OS alternatives will simply keep nibbling away.

**Comment by murphomatic** [comment]
> We too use on prem minio storage and would not like to see this changed. It makes sense to pay for the features that NX provides as a service, but for us to pay for using a shared fs cache to setup ourselves as well... What is that about?
> 
> We've been (and myself personall) very happy with NX all through the years, but changes like this makes me rethink what the direction of the whole organisation is moving forward. As others have stated, what will be the next paywall?

Just wanted to echo the sentiment in the post above by @johannesnormannjensen.  

We too have been using Nx for a few years (with varying degrees of satisfaction, but pretty happy with it for the most-part), and while I can understand a pay-to-play model for consumers who wish to utilize Nx's cloud-based infrastructure for caching/analytics/build process, etc., I'm hard-pressed to understand hobbling framework (and requiring a paid-package to unhobble it) for those who have no ties to Nx cloud infrastructure.  In our particular situation, we are fairly tightly-controlled with respect to data egress, and we use nx-remotecache-azure so that our cacheable assets never leave the confines of our corporate network.

Ive hemmed and hawed a bit about whether or not we'd just bite the bullet and "pay da man," but after fielding some bizarre build issues today that appear be related to #28599 and #28380, my confidence in the Nx tooling has taken a bit of a nose-dive.  Properly sequenced builds within complex dependency chains is a core, if not -the- core, reason we chose Nx in the first place, and coming to find this broken - well ... let's just say that the corporate credit card swiftly moved back into safe confines of my wallet.

And in continuing to emphasize @johannesnormannjensen's post above, as well as @Elyahou's similar point - what is the next thing that will become pay-to-play, even if opt to we forgo computational caching for now?  The ability to make implicit dependencies between projects?  Limits on the number of parallel task runners?  Maximum number of projects in an Nx workspace?

This approach that does not seem like a sensible/logical "pay for what you use" model, but instead of more a "if we perceive you to be a big, evil, greedy corporation - we're going to work you like a mule."  This kind of logic is made even more apparent by the attempt to assuage the angst in this thread by making self-hosted remote-caching with powerpack "free for small orgs."  What difference does it make to Nx how big the org is if they're self-hosting their own cache?  How does Nx define a "small org," and what will prevent that definition from changing tomorrow?

There are pay-to-play models that would make more sense here, in my opinion:
* Recurring subscription for using Nx Cloud, and the features it provides
* Recurring subscription for Enterprise-grade support with varying degrees of SLAs
* Even a licensing fee for major versions of Nx where consumers could opt to pay once and then stay with a particular major version for a longer period of time.  Like we might decide to stay on v19 for the next 3 years ... I don't want my org to hemorrhage nearly $30k simply because .... "reasons."

Perhaps the time has come to start planning our move away from Nx.

**Comment by Elyahou** [comment]
> Thank you for engaging with us!
> 
> Anyone who has spoken with us in person at conferences knows that the Nx team, myself included, is always happy to discuss what's in progress and what's coming next. However, we want to do a much better job of sharing this in a more public way, so everyone interested can stay informed and have an opportunity to share feedback early on.
> 
> To achieve this, we're making a few changes:
> 
> 1. Publishing a 1-Year Roadmap: We're bringing back roadmaps and revamping the process. This roadmap will outline all major developments for the year, offering a chance to see what's in progress, ask questions, and share concerns. The 2025 roadmap will be published soon.
> 2. A More Formal RFC Process: For larger API changes (e.g., Crystal, Powerpack are past examples of such changes), we'll pilot a more structured RFC process, providing an opportunity for you to see upcoming changes early and contribute directly to the discussion. We commit to implementing major changes based on the resolutions of these RFCs, ensuring that your input guides our direction.
> 3. Hosting Regular Office Hours: We'll host regular office hours, providing a chance for anyone to have direct access to the team.
> 4. Improving the Nx Champions Program: The Nx Champions program connects 28 prominent community members with the Nx team for feedback and idea exchange. We haven't been using this program as effectively as we'd like. We're going to take action to improve the program such that there's more collaboration, early feedback gathering but also getting their help to spread big ongoing developments better in the community.
> 
> I'm personally committed to engaging with the community. If you have questions or concerns, feel free to DM me on [Twitter](https://x.com/victorsavkin), and I'll be happy to have a video call with you. You can also engage with Juri in the [same fashion](https://x.com/juristr).
> 
> Discord is also a great place to engage at.
> 
> I'm sorry we didn't do a good enough job of communicating with the community. We are doing our best to address it.
> 
> If you have any questions about these updates, please engage with us!

While the proposed changes are a step in the right direction, they don’t address the immediate issue: this decision was made without community input, and the feedback so far has been overwhelmingly negative, as shown in this issue...

![Image](https://github.com/user-attachments/assets/d55e2f91-5917-440d-9cc4-55ea98817397)

These changes are planned for future versions of Nx, and since it’s unclear when the migration to Rust will happen, there’s still time to reconsider. Deprecating the API before Rust is fully implemented creates unnecessary friction. To show good faith and rebuild trust, this decision should be revisited with the community’s input.

**Comment by maxisam** [comment]
It is a shame I really like NX, but I guess they have to somehow get money from NX. I totally understand that but $25 per users per month is a very expensive price. 

Github copilot only cost like $20 for business and it provides more value. 
Docker team is like $5 -$10 for a small team (< 100 ppl), and it provides lots of resource as well.

If we don't use nx cloud resource, I really can't justify $25 for a remote cache feature that actually using our own resource. 
I think it will be very hard to sell this to a lot of smaller companies that is not in CA or big cities.

I think it is probably impossible to ask NX team to change their mind on paying, but I think could you guys at least provide something cheaper for smaller teams like ppl < 100 or 50? I think docker's pricing is a good one

**Comment by maxisam** [comment]
Honestly, I understand why NX decided to put the remote cache feature behind a paywall. Building and maintaining NX takes a lot of time and resources. I just hope the NX team can find a sustainable way to keep moving forward.

As an open-source developer, I recognize it’s unrealistic to expect everything for free indefinitely. I personally don’t charge for my open-source projects since it's a passion project for me. I even convinced my boss to let me support open-source contributions for the tools we use in our company as a small way to give back to the community. However, I believe there's a difference when people are working full-time to support these projects. Let's face it, if NX can't keep going forward, I think it is still a bad outcome for everyone. There are so many open source projects died already. 

I’m grateful that NX provides us with a free Power Pack since we’re a small team (under 15 people).

🙏 I hope the community can come together to find a balanced solution.

**Comment by wall-street-dev** [comment]
Dear @vsavkin & @jeffbcross,
You guys are doing it wrong and you know it.
Yes, you'll get more money, and if that's your goal then it's fine. But as an OSS project, your credibility and Nx's credibility are being undermined by these types of decisions.

Godspeed.

**Comment by o629625** [comment]
This post did not age well https://www.linkedin.com/posts/zackderose_we-were-intentional-with-nx-to-allow-other-activity-7069448865090859008-vnLI 

_[Zachary DeRose](https://www.linkedin.com/in/zackderose?trk=public_post_feed-actor-name)
**Engineering Manager and Architect at Nx**

💯 We were intentional with Nx to allow other task runners that implement their own distributed caching!

But caching can be hard. Nx Cloud is the first-party solution here (checkout https://[nx.app](http://nx.app/?trk=public_post-text)) and we're investing our engineering resources into making this an excellent experience for our users.

Nx Cloud is free up to 300 CI runs a month (all non-CI runs are free) and there's a special unlimited tier for all open-source projects, and comes with some nice features like linkable task logs, Version Control System integrations for Github/GitLab/BitBucket, and support for Distributed Task Execution!

You can try it now simply by running `nx connect` in your Nx Workspace 😎_


----

NX team must reconsider removing existing open source functionality they **intentionally** offered. The disruption you are causing existing teams using custom runners will likely cost more then net revenue from placing it behind a pay wall.

**Comment by ytchenak** [comment]
My CI/CD pipeline runs concurrent NX processes in the same workspace (e.g., build and lint). Starting with version 19.8.*, I've started seeing occasional SQLite errors, which suggests the metadata SQLite database might be corrupted by concurrent writes from multiple NX processes.

Will this scenario (concurrent NX processes in the same workspace) be supported in future versions, or do I need to purchase the Powerpack to use the remote FS cache?

**Comment by alexhanga** [comment]
We have a custom runner that calls our own API with security checks and authorization, which also have some checks for a cache before storing it to S3. So the developers don't have a real access to Amazon infrastructure and to S3 itself.

Powerpack doesn't meet our needs in security and with restricted access to Amazon infrastructure. So basically with this change, we have only three options:

1. Remove caching at all
2. Stay indefinitely on Nx 20
3. Change our security policy to be less strict and give access to Amazon infrastructure just to be able to use cache. And pay for using remote cache.

I don't think that this options are good.

**Comment by JoeRJohnson11** [comment]
@o629625 Can we have a call to discuss your situation? If you email [powerpack-support@nrwl.io](mailto:powerpack-support@nrwl.io) we'll find a time to talk.

**Comment by ytchenak** [comment]
> [@ytchenak](https://github.com/ytchenak) " My CI/CD pipeline runs concurrent NX processes in the same workspace (e.g., build and lint). Starting with version 19.8.*, I've started seeing occasional SQLite errors" It's a bug that should be fixed. If you upgrade to latest, the issue should go away.
> 
@vsavkin, I’m glad to hear that there’s no additional cost for running concurrent jobs, but unfortunately, the issue persists even in the latest version (20.1.2). Please refer to this comment for more details: https://github.com/nrwl/nx/issues/28772#issuecomment-2485635108.  

This issue is preventing us from upgrading to the latest NX versions. For now, I am forced to use the last NX version that does not rely on SQLite.

**Comment by vsavkin** [comment]
Folks,

I'd like to post a quick update on the process initiatives that should give a lot more visibility into the roadmap, how features are developed etc:

1. We published 2025 roadmap: https://github.com/nrwl/nx/discussions/28731 
2. We are trying a new RFC process. This is the first BIG feature: https://github.com/nrwl/nx/discussions/29025  I also created a video showing what we are trying to do https://www.youtube.com/watch?v=-gezeX9zxuM
3. We started hosting regular office hours on Discord. More details to come soon.

Let us know what we can do to tweak the process changes (especially the RFC) if they are useful or not.

**Comment by TriPSs** [comment]
It could be just me (and I really hope it is), but we are now a couple of weeks further along, and although a new RFC process has been introduced, we still don't have any clear answers regarding the custom runners?

**Comment by el-davo** [comment]
@TriPSs I have to agree. The communication and side stepping here is not great. Please answer the following questions so we can make a decision on our future of using nx

- Will the removal of custom runners have an RFC?
- Are NX going ahead with putting private caching behind a paywall?

Any other response is redundant as far as my organization is concerned

**Comment by JoeRJohnson11** [comment]
@el-davo & @TriPSs Would you be willing to have a call to talk about your situation? If you write into powerpack-support@nrwl.io, we'll find a time to talk.

**Comment by joshlartz** [comment]
https://github.com/nrwl/nx/discussions/28360

**Comment by bahizi** [comment]
Hi @vsavkin first of all, thanks for all the work you and the team have been doing over the years to support nx. I understand it takes time and effort to build a powerful tool like this, open source or not, so I also understand the need for the powerpack commercial products, but can we reconsider the vendor lock-in aspect of this change? 

Remote caching is crucial for NX in a CI/CD environments that use docker containers to run jobs (which I think is how most folks do CI/CD nowadays). Frankly without remote cache and therefore no `Fast CI` (a previously key out-the-box attractive nx feature, along with nx affected), nx (and monorepos in general) becomes somewhat impractical and also goes against some of the NX key pros that were listed in this post a while ago: https://nx.dev/concepts/turbo-and-nx, taking the overall nx project on the path to an arguably freemium product as well.


Again, we're not against the powerpack products, but please leave options for custom remote cache implementations.

**Comment by JoeRJohnson11** [comment]
@bahizi would you mind having a call with us so we can better understand your situation?
If you write into powerpack-support@nrwl.io we'll find a time to talk.

**Comment by TriPSs** [comment]
Hi @JoeRJohnson11,

In our call a couple of weeks ago, I asked if you could provide an update on this thread. Since that hasn’t happened yet, I’ll take the liberty to share my understanding of the situation. If anything is incorrect, please feel free to clarify.

Key points discussed:
- The Rust migration is here to stay, which includes the deprecation and removal of custom runners.
- Users can apply for a "free" Powerpack license to continue using their own AWS/GCP buckets for remote caching.
- Nx is open to working with users who have other custom implementations to find potential solutions. (However, we didn’t address the broader perception of this change, or the concern that no feature is guaranteed to remain outside the license wall.)

As I mentioned during the call, I will not be applying for a "free" license to continue using something I’m already using. Additionally, there’s the concern that the license could become perpetual for a specific version, which I understand to mean that while it’s free now, it might not remain so with future Nx versions.

My conclusions:
- I’m no longer worried about the Rust migration, as Nx assured us that they will work with plugin maintainers to ensure compatibility. Based on how the migration from our Playwright plugin to Nx’s version was handled, I’m confident this will go smoothly.
- However, I’m still concerned about the removal of custom runners. I’m uncertain about the path forward—whether I’ll continue using Nx until the removal, attempt a workaround, or move away from Nx entirely.

Answers to our key questions:
> Will the removal of custom runners have an RFC before implementation?

No.

> Is Nx moving forward with putting private remote caching behind a paywall?

Yes.

PS: I also noticed that Vercel made their caching free ([source](https://vercel.com/changelog/free-vercel-remote-cache)). It even states that Nx users as of version 19.7+ may need to pay a license fee to Nx to continue using Vercel Remote Cache.

Let me know if I’ve missed or misunderstood anything!

**Comment by wvanderdeijl** [comment]
As the maintainer of [nx-remotecache-gcs](https://github.com/wvanderdeijl/nx-remotecache-gcs), an Nx plugin that enables Google Cloud Storage for remote caching, I’m disappointed by this decision. To safeguard our organization, we’ve applied for a free Powerpack license, but it’s only valid for one year, requiring renewal annually.

Is there any assurance that these licenses will continue to be free in the future?

While we’ll observe how this unfolds in the near term, the gradual shift toward paywalling remote caching doesn’t sit well with us. Given the current direction, we’re actively exploring alternatives to Nx.

**Comment by o629625** [comment]
In our organization, I have recently come across custom remote cache implementations that use a REST API to interact with the cache store, because it it not possible to connect directly to the S3 bucket from CI. 

It doesn't look like Powerpack will work here. This is another example of the variety of ways the NX community have used custom runners and if NX moves forward with removing them, many may have no choice but to migrate to a different monorepo solution.

**Comment by vsavkin** [comment]
Folks, 

We are introducing a new API: https://github.com/nrwl/nx/discussions/29637

It handles most of the functionality that custom tasks runners were used for, but in a much clearer way and without impending our ability to evolve Nx. 

Please take a look and provide feedback.

Thank you!

**Comment by Jordan-Hall** [comment]
> Folks,
> 
> We are introducing a new API: [#29637](https://github.com/nrwl/nx/discussions/29637)
> 
> It handles most of the functionality that custom tasks runners were used for, but in a much clearer way and without impending our ability to evolve Nx.
> 
> Please take a look and provide feedback.
> 
> Thank you!

Thank you, but I'm not sure if the two is linked. It appears most here saying custom runner for a wide range of issues. I'll leave a comment anywsy, but looks like custom caching harder/may not be as good here :/ 

I still think and rust needs a API around caching that people can hook into as well.

**Comment by vsavkin** [comment]
@wvanderdeijl
> Is there any assurance that these licenses will continue to be free in the future?
Yes. Your free license will always be free. Also, non-free licenses are perpetual, so they will continue working after expiring with version restrictions.

**Comment by med8bra** [comment]
It seems nx team is fully committed on making self-hosted remote cache a subscription service, and to host your cache you will need to go through a paywall. Not sure what was the basis of this product decision, but I believe nx has more potential than just to be a payme for cache product. 

[Turborepo documents their Remote Caching features](https://turbo.build/repo/docs/core-concepts/remote-caching), documented the API specification, and allowed the community to implement [self hosted cache solutions](https://turbo.build/repo/docs/core-concepts/remote-caching#self-hosting).

**Comment by Playjasb2** [comment]
I understand that the Nx team needs funding to sustain themselves, but putting the custom remote caching feature behind a paywall is disappointing. I was an advocate for using Nx in my company, and we are deeply invested in the way that Nx works in our project, and I recently got a development server and I intended to use that for our remote caching server, and now this news hit.

My team is too used to the way that Nx works in our project, and it's too expensive in time and effort to move to another monorepo framework, as I had documented Nx use cases in our own GitHub wiki and it's used in our GitHub Actions pipeline. Plus, the company I'm in is not a small business, so we won't qualify for a free license.

We are effectively vendor-locked, which I didn't exactly see coming, since it's a FOSS, and I thought that given the reputation Nx had with the community, they wouldn't do something like backing away from a prior promise of a free remote caching solution. There are a lot of people that bought into Nx for the free remote caching option, and now they're stuck.

Right now, I'm trying to understand if I should use Nx in the next project I'll be working on. Issue is that I spent quite a lot of time learning Nx, how it works, and how to resolve the various quirks that it has, that jumping to another monorepo framework would be another issue, since I have to start from the very beginning again. Plus I got used to Nx that I'm not sure if any other frameworks has the same equivalent feature set.

But this may not have been the case for anyone else if this was not an advertised feature from the start, as others would've just used free remote caching features offered by Rush, Bazel, Turborepo, etc. Cause at that point, some people wouldn't have given Nx a chance there if they prioritized free remote caching.

But given the situation, it feels like I would be more inclined to use Nx again considering the time invested in being familiar with it. But I'd have to lock myself to v20 to keep the promised free feature...

**Comment by mleithaug** [comment]
I hope I have misunderstood something but this just all seems so disingenuous. All the comments in this thread talk about the removal of custom runners people used to self-host remote caching, yet the only responses we get from nx is about everything else.  
<br/>

>Folks,
>
> We are introducing a new API: https://github.com/nrwl/nx/discussions/29637
> 
> It handles most of the functionality that custom tasks runners were used for, but in a much clearer way and without impending our ability to evolve Nx.
> 
> Please take a look and provide feedback.

This is unrelated to custom remote caching and we're event told this:
> This API offers a different set of capabilities as detailed above. Discussion about caching can be done elsewhere.

<br/>

People asking simple yes/no questions like: 
>Will the removal of custom runners have an RFC before being implemented?
>Are NX going ahead with putting private remote caching behind a paywall?

Are asked for private communication instead.  This feels really shady.
<br/>

They called out some bad behavior from one commenter offering to fork nx to keep custom runners working. It might be true, and it seems correct to not accept the described behavior. But again why post that and not answer about caching.

Then there is even a PR to alter the documentation for migrating to turborepo. I couldn't find find anything misleading with it  but at this point I'm all out of trust and just assume I couldn't find it.
 
We get that you need to pay the bills. But why not just say that. Why coun't you just say that. Why blame it on a Rust rewrite and throw us red herrings about irrelevant PR,  roadmaps and make us read through threads and PRs, and videos on twitter, maybe there is a discord thread. 

It actually seems like their current plan is quite reasonable. Smaller teams (less than 30 devs) get free remote caching but not the other upcoming features of Powerpack. If nx started with that offering it might have been fine, but they did not .

It seems I will be able to use nx at work past v21 so I have no technical issue to report here, but my trust in nx is severly diminished and I will more actively look for alternatives. But I guess we're too locked in at this point, and it feels like nx knows that too well.

**Comment by wall-street-dev** [comment]
> My advice, NX team should continue with a solution that allows users to self-host a remote cache AND also offer a more "white glove" Powerpack solution.
> 
> Market the benefits of Powerpack moving forward to drive adoption. After all, Powerpack is more then just a remote caching solution? And features will continue to be added? Making it more compelling for teams to onboard.
> 
> This way the existing users of NX that already have a self hosted cache are not impacted, community trust in the product and NX leadership is maintained, and the NX team continues with the Powerpack strategy.
> 
> Win win for everyone. Just do it! It is the right thing to do.

I have a better advice to the Nx Team:

![Image](https://github.com/user-attachments/assets/d854b0d6-bda5-4f08-b45a-a9d3ae62a2b5)

**Comment by vsavkin** [comment]
Folks,

We've been in touch with many teams over the past few months. We hear you, we will be announcing a free self-hosted cache solution soon. We're validating it with our champions and other organizations. Early feedback has been very positive, give us some time to make sure we get this right. More details including a community RFC are coming soon. Rest assured Nx 21 will not be released until this is all in place.

**Comment by Jordan-Hall** [comment]
> Folks,
> 
> We've been in touch with many teams over the past few months. We hear you, we will be announcing a free self-hosted cache solution soon. We're validating it with our champions and other organizations. Early feedback has been very positive, give us some time to make sure we get this right. More details including a community RFC are coming soon. Rest assured Nx 21 will not be released until this is all in place.

Thank you!!!! I had a feeling this was coming but the deadly silents been concerning. While we here i think more empathy on what powerpacks overall provides (and what community cant add to it) also maybe change it so you can get set powerpack rather than them all?

**Comment by o629625** [comment]
> Folks,
> 
> We've been in touch with many teams over the past few months. We hear you, we will be announcing a free self-hosted cache solution soon. We're validating it with our champions and other organizations. Early feedback has been very positive, give us some time to make sure we get this right. More details including a community RFC are coming soon. Rest assured Nx 21 will not be released until this is all in place.

Excellent news!! Thanks @vsavkin and team!

**Comment by JelleBruisten** [comment]
> Folks,
> 
> We've been in touch with many teams over the past few months. We hear you, we will be announcing a free self-hosted cache solution soon. We're validating it with our champions and other organizations. Early feedback has been very positive, give us some time to make sure we get this right. More details including a community RFC are coming soon. Rest assured Nx 21 will not be released until this is all in place.

Great to hear, happy this will get resolved. Looking forward to the RFC

**Comment by mleithaug** [comment]
> Folks,
> 
> We've been in touch with many teams over the past few months. We hear you, we will be announcing a free self-hosted cache solution soon. We're validating it with our champions and other organizations. Early feedback has been very positive, give us some time to make sure we get this right. More details including a community RFC are coming soon. Rest assured Nx 21 will not be released until this is all in place.

That's such a relief! Thank you!

**Comment by AlonMiz** [comment]
> Folks,
> 
> We've been in touch with many teams over the past few months. We hear you, we will be announcing a free self-hosted cache solution soon. We're validating it with our champions and other organizations. Early feedback has been very positive, give us some time to make sure we get this right. More details including a community RFC are coming soon. Rest assured Nx 21 will not be released until this is all in place.

These are very good news for us and i think for the future of nx
Thanks @vsavkin

**Comment by xalechez** [comment]
Thank you @vsavkin and the team! I greatly appreciate that decision which will be a huge win for everyone

**Comment by el-davo** [comment]
While this is obviously great news I would caution everyone to be vigilant in the future of companies trying things like this. I think we can all agree that the initial rollout was trying to use some very questionable methods of extracting more money from users.

I will wait for the final RFC before making a decision to stay with NX or not.

**Comment by intermundos** [comment]
Damn, just wanted to give it a run...

**Comment by tonyganchev** [comment]
Whatever you do from this point on, you did try the bait and switch and the community will never trust you again. My team's projects are all moving away from nx. Congrats and good luck with the IPO!

**Comment by suhas-coursera** [comment]
> Whatever you do from this point on, you did try the bait and switch and the community will never trust you again. My team's projects are all moving away from nx. Congrats and good luck with the IPO!

I'm curious to know where you're moving to. What other options do we have that compete with NX?

**Comment by maxisam** [comment]
> Whatever you do from this point on, you did try the bait and switch and the community will never trust you again. My team's projects are all moving away from nx. Congrats and good luck with the IPO!

ah....where is the news about IPO, do I miss anything? I would love to buy some :D

Seriously, we have to face reality, free open source is noble but it is hard for maintainers. 

So just give them a break, at least they tried to take care smaller teams / org. 

Anyway, for the record, thank you @vsavkin for making it free!

**Comment by vsavkin** [comment]
Folks,

 We changed our plans. We’re implementing an API to allow self-hosted caching. Everything will be in place before Nx 21, ensuring a smooth transition for everyone. Custom self-hosted caching will possible in all versions of Nx.

 Please read this blog post explaining what happened, our new solution, and our OSS commitment to the Nx community:
[Custom Task Runners and Self-Hosted Caching Changes](https://nx.dev/blog/custom-runners-and-self-hosted-caching)

Thank you so much for engaging with us!
Victor

**Comment by Jordan-Hall** [comment]
> Folks,
> 
>  We changed our plans. We’re implementing an API to allow self-hosted caching. Everything will be in place before Nx 21, ensuring a smooth transition for everyone. Custom self-hosted caching will possible in all versions of Nx.
> 
>  Please read this blog post explaining what happened, our new solution, and our OSS commitment to the Nx community: [Custom Task Runners and Self-Hosted Caching Changes](https://nx.dev/blog/custom-runners-and-self-hosted-caching)
> 
> Thank you so much for engaging with us! Victor

 Thank you! That's an absolutely amazing announcement. As one of the more vocal people advocating for this, I appreciate that you've gone above and beyond. No one expected a refund, especially since we didn’t have to pay for the packages initially. Even if they were once free (e.g., S3, Azure, etc.), as long as we still had the ability to use them, we were happy to pay a premium for the maintenance of those packages.

**Comment by vsavkin** [comment]
@Jordan-Hall Thank you, Jordan. I appreciate your engagement. The community is strong when people like you care!

**Comment by rathpc** [comment]
> Folks,
> 
>  We changed our plans. We’re implementing an API to allow self-hosted caching. Everything will be in place before Nx 21, ensuring a smooth transition for everyone. Custom self-hosted caching will possible in all versions of Nx.
> 
>  Please read this blog post explaining what happened, our new solution, and our OSS commitment to the Nx community:
> [Custom Task Runners and Self-Hosted Caching Changes](https://nx.dev/blog/custom-runners-and-self-hosted-caching)
> 
> Thank you so much for engaging with us!
> Victor

This is incredible news and I am so happy that you all at Nrwl decided to take the feedback and response from the community to heart and change your initial plans. It is very appreciated and I am sure it will help regain the trust of many who decided to go in a different direction after the first announcement. Thanks again!

**Comment by AlonMiz** [comment]
> Folks,
> 
>  We changed our plans. We’re implementing an API to allow self-hosted caching. Everything will be in place before Nx 21, ensuring a smooth transition for everyone. Custom self-hosted caching will possible in all versions of Nx.
> 
>  Please read this blog post explaining what happened, our new solution, and our OSS commitment to the Nx community: [Custom Task Runners and Self-Hosted Caching Changes](https://nx.dev/blog/custom-runners-and-self-hosted-caching)
> 
> Thank you so much for engaging with us! Victor

This is amazing news! It was way beyond what I expected. Huge thanks, Viktor, for this incredible contribution to the OSS world and for listening to the community

**Comment by AlonMiz** [comment]
We ran some benchmarks measuring cold start times for tasks (in seconds) locally. Looks like moving away from custom runners to a faster alternative could yield significant improvements. Definitely seems worth the switch!

<img width="657" alt="Image" src="https://github.com/user-attachments/assets/a7755ff8-d9b2-4edf-9055-d027da7e097f" />

**Comment by juristr** [comment]
@AlonMiz what is Legacy Cache and Modern Cache for you here in this diagram?

**Comment by tonyganchev** [comment]
> > Whatever you do from this point on, you did try the bait and switch and the community will never trust you again. My team's projects are all moving away from nx. Congrats and good luck with the IPO!
> 
> ah....where is the news about IPO, do I miss anything? I would love to buy some :D

I have not heard but usually this is the expected next step for such projects after they start trying to monetize.

> 
> Seriously, we have to face reality, free open source is noble but it is hard for maintainers.

True, that's why there are huge foundations and companies backing the highest quality OSS dev tools - talking about Apache projects, or Spring. None of them require payment or have ever cut features that are already used by the community, have unparallel quality, and extensive documentation.

> So just give them a break, at least they tried to take care smaller teams / org.

Talking about smaller teams: these are the type of teams that ask questions on the nrwl discord and have progressively not received any responses in the last two years. Same people that are reporting issues that stay stale for months (even for things that are obvious regressions on the part of nrwl), only to get these bugs auto-closed. So, I am not sure what amount of support they are gonna get if they qualify for a free premium subscription.

> Anyway, for the record, thank you [@vsavkin](https://github.com/vsavkin) for making it free!

He's not making anything new free. He simply decided not to take it away from you.

**Comment by tonyganchev** [comment]
> > Whatever you do from this point on, you did try the bait and switch and the community will never trust you again. My team's projects are all moving away from nx. Congrats and good luck with the IPO!
> 
> Agree [@tonyganchev](https://github.com/tonyganchev), I'm curious to know where you're moving to. Also what all options do we have that compete with NX?

@suhas-coursera , my company is already heavily invested in Bazel for an actual real monorepo (not just UI or a single product). Back in the day, Bazel was hard to integrate with frontend tools but more and more this is not the case. It gives us build task orchestration, distributed caching and remote execution (although to me the last's need is questionable). Moreover, based on Bazel's architecture we have something, Nx cannot guarantee with its caching and that is build reproducibility.

**Comment by Exelord** [comment]
Licensing, Cache Service OpenAPI... why not just let people decide if they want convenience, support and first party maintenance OR use whatever custom solution they need?

As simple solution as removing the bloatware from the code.

Liberate Remote Cache https://github.com/nrwl/nx/pull/30604

**Comment by vsavkin** [comment]
Quick update:
We worked with the community to implement the self-hosted cache RFC. The Nx 20.8.0 release contains the change.
You can find the documentation here: https://nx.dev/recipes/running-tasks/self-hosted-caching#build-your-own-caching-server 
Thank you to everyone who contributed to this effort!

If you haven’t had a chance to read it, please check out this post: https://nx.dev/blog/custom-runners-and-self-hosted-caching
I want to emphasize a crucial point: “Our commitment is that no major open-source feature or API will be replaced with a non-open-source alternative.”
If you have any questions or concerns regarding this, feel free to reach out to me via Twitter or Bluesky. I’ll be happy to jump on a call with you.

**Comment by RichardLindhout** [comment]
The activation process does not work though?

```
 NX   No Nx keys were found for this organization. Please visit https://cloud.nx.app/powerpack to purchase or renew your key.

Pass --verbose to see the stacktrace.

✖ Initializing @nx/gcs-cache...


 NX   Failed to initialize @nx/gcs-cache
```
There is nowhere a link to get an activation key?
